pmd-core/src/main/java/net/sourceforge/pmd/AbstractPropertyDescriptorFactory.java

```diff
@@ -28,30 +30,42 @@
  */
 public abstract class AbstractPropertyDescriptorFactory<T> implements PropertyDescriptorFactory<T> {
 
-    protected static final Map<PropertyDescriptorField, Boolean> CORE_FIELD_TYPES_BY_KEY
-        = CollectionUtil.mapFrom(new PropertyDescriptorField[] {NAME, DESCRIPTION, DEFAULT_VALUE, DELIMITER},
-                                 new Boolean[] {true, true, true, false});
+
+
+
+    static {
+        System.err.println(CORE_FIELD_TYPES_BY_KEY);
```

Comment:
The cycle need not be a 2-cycle, it could be rather long and circuitous.  Try putting a `main` method on `IntegerProperty`, and run it using the `-verbose:class` JVM argument.  You should see the order of classloading with `IntegerProperty` coming first, and hopefully `PropertyDescriptorUtil` eventually.  I think the static initialization likely correlates well with this loading.  If so, it seems reasonable that a _portion_ of _a_ cycle between will be included in the classes listed between them, including perhaps some unrelated classes not part of the cycle path.  You could also try this starting on the `PropertyDescriptorUtil` and see another part of the cycle, but I think that's the part you want to keep.

It makes sense for the `IntegerProperty` to appear initialized afterward, it's just not _fully_ initialized in the middle of the cycle.  Once the JVM finishes initializing `PropertyDescriptorUtil` it will continue with the initialization of `IntegerProperty`.

This [SO answer](https://stackoverflow.com/questions/30142611/recursive-initialization-static-initializer-not-getting-called-when-i-access-c) is more directly related.

See [JLS 12.4](https://docs.oracle.com/javase/specs/jls/se8/html/jls-12.html#jls-12.4) for the initialization process.  This wording from 12.4.1 speaks to the fact that `null` (the default value) is returned in the case you're hitting.

> The fact that initialization code is unrestricted allows examples to be constructed where the value of a class variable can be observed when it still has its initial default value, before its initializing expression is evaluated, but such examples are rare in practice. (Such examples can be also constructed for instance variable initialization (§12.5).) The full power of the Java programming language is available in these initializers; programmers must exercise some care. This power places an extra burden on code generators, but this burden would arise in any case because the Java programming language is concurrent (§12.4.2).